File: ios/Anemone SDK/general/json/JSONTransformer.swift
//
//  JSONTransformer.swift
//  Anemone SDK
//
//  Created by Manuel Gonzalez Villegas on 5/12/16.
//  Copyright © 2016 Syltek Solutions S.L. All rights reserved.
//

import Foundation

public class JSONTransformer {
    public var rootKey: String?

    public init() { }

    public func transform<T: JSONMappable>(data: Data) -> T? {
        do {
            guard let json = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] else {
                return nil
            }

            let jsonObject: JSONObject
            if let rootKey = rootKey {
                jsonObject = try JSONObject(data: json).getJSONObject(rootKey)
            } else {
                jsonObject = JSONObject(data: json)
            }
            return instantiate(json: jsonObject)
        } catch {
            Log.w("JSONTransformer", "Can not transform object of type \(T.self) because \(error)")
            return nil
        }
    }

    public func transform<T: JSONMappable>(data: Data) -> [T]? {
        do {
            let json = try JSONSerialization.jsonObject(with: data, options: [])

            let jsonArray: [[String: Any]]?
            if let rootKey = rootKey {
                jsonArray = (json as? [String: Any])?[rootKey] as? [[String: Any]]
            } else {
                jsonArray = json as? [[String: Any]]
            }

            return jsonArray?.compactMap { instantiate(json: JSONObject(data: $0)) }
        } catch {
            Log.w("JSONTransformer", "Can not transform array of type \(T.self) because \(error)")
            return nil
        }
    }

    public func map<T: JSONMappable>(_ data: Data) -> Promise<T> {
        Promise(executeInBackground: true) { fulfill, reject in
            if let object: T = self.transform(data: data) {
                fulfill(object)
            } else {
                reject(AnemoneException.notMappable)
            }
        }
    }

    public func map<T: JSONMappable>(_ data: Data) -> Promise<[T]> {
        Promise(executeInBackground: true) { fulfill, reject in
            if let objects: [T] = self.transform(data: data) {
                fulfill(objects)
            } else {
                reject(AnemoneException.notMappable)
            }
        }
    }

    @inline(__always)
    private func instantiate<T: JSONMappable>(json: JSONObject) -> T? {
        do {
            return try T(json: json)
        } catch {
            Log.w("JSONTransformer", "Can not instantiate object of type \(T.self) because \(error)")
            return nil
        }
    }
}

File: android/app/src/main/java/com/playtomic/test/anemone_sdk/general/json/JSONTransformer.kt
package com.anemonesdk.general.json

import com.anemonesdk.general.exception.AnemoneException
import com.playtomic.foundation.logger.Log
import com.playtomic.foundation.promise.Promise
import java.lang.reflect.Constructor

/**
 * Created by mgonzalez on 21/12/16.
 */

class JSONTransformer<T : JSONMappable>(val clazz: Class<T>) {
    var constructor: Constructor<T>? = null

    init {
        try {
            constructor = clazz.getDeclaredConstructor(JSONObject::class.java)
        } catch (t: Exception) {
            Log.e("JSONTransformer", "Object $clazz does not implement JSONMappable")
        }
    }

    var rootKey: String? = null

    fun transformObject(data: ByteArray): T? {
        try {
            val stringObject = String(data)
            val orgJson = org.json.JSONObject(stringObject)
            var json = JSONObject(orgJson)

            rootKey?.let { json = json.getJSONObject(it) }

            return instantiate(json)
        } catch (je: Exception) {
            Log.w("JSONTransformer", "Can not transform object of type $clazz because $je")
            return null
        }
    }

    fun transformArray(data: ByteArray): List<T>? {
        return try {
            val stringArray = String(data)
            val jsonArray: JSONArray = if (rootKey != null) {
                val orgJson = org.json.JSONObject(stringArray)
                JSONArray(orgJson.getJSONArray(rootKey ?: ""))
            } else {
                val array = org.json.JSONArray(stringArray)
                JSONArray(array)
            }
            jsonArray.flatMap(this::instantiate)
        } catch (je: Exception) {
            Log.w("JSONTransformer", "Can not transform array of type $clazz because $je")
            null
        }
    }

    fun mapObject(data: ByteArray): Promise<T> =
        Promise(executeInBackground = true) { fulfill, reject ->
            val `object` = transformObject(data)
            if (`object` != null) {
                fulfill.invoke(`object`)
            } else {
                reject.invoke(AnemoneException.notMappable)
            }
        }

    fun mapArray(data: ByteArray): Promise<List<T>> =
        Promise(executeInBackground = true) { fulfill, reject ->
            val objects = transformArray(data)
            if (objects != null) {
                fulfill.invoke(objects)
            } else {
                reject.invoke(AnemoneException.notMappable)
            }
        }

    private fun instantiate(json: JSONObject): T? =
        try {
            constructor?.newInstance(json)
        } catch (e: Exception) {
            Log.w("JSONTransformer", "Can not instantiate object of type $clazz because ${e.cause}")
            null
        }
}
